addons/t9n/models/resource.py

```diff
@@ -14,3 +19,84 @@ class Resource(models.Model):
     project_id = fields.Many2one(
         comodel_name="t9n.project",
     )
+
+    _sql_constraints = [
+        (
+            "file_name_project_id_unique",
+            "unique(file_name, project_id)",
+            "A file with the same name already exists in the same project!"
+        ),
+    ]
+
+    def _decode_resource_file(self, resource_file):
+        try:
+            file_content = base64.b64decode(resource_file).decode()
+            po_obj = polib.pofile(file_content)
+        except (IOError, UnicodeDecodeError):
+            return None
+        return po_obj
+
+    def _get_message_from_file(self, resource_file):
+        for entry in resource_file:
+            references_list = [reference[0] + (f":{reference[1]}" if reference[1] else "")
+                               for reference in entry.occurrences]
+            entry.references = references_list
+
+        messages_to_create = [{"body": entry.msgid, "context": entry.msgctxt, "translator_comments": entry.tcomment,
+                               "extracted_comments": entry.comment, "references": "\n".join(entry.references)} for entry in resource_file]
+        return messages_to_create
```

Comment:
one line, one key. It's clearer this way, and will allow future additions to the list without diffing the line
```suggestion
        return [
            {
                "body": entry.msgid,
                "context": entry.msgctxt,
                "extracted_comments": entry.comment,
                "references": "\n".join(entry.references),
                "translator_comments": entry.tcomment,
            }
            for entry in resource_file
        ]
```